sync/src/block/extension.rs

```diff
@@ -743,20 +743,22 @@ impl Extension {
     fn on_header_response(&mut self, from: &NodeId, headers: &[Header]) {
         ctrace!(SYNC, "Received header response from({}) with length({})", from, headers.len());
         match self.state {
-            State::SnapshotHeader(..) => {
-                for header in headers {
+            State::SnapshotHeader(hash, _) => match headers {
+                [header] if header.hash() == hash => {
                     match self.client.import_bootstrap_header(&header) {
-                        Err(BlockImportError::Import(ImportError::AlreadyInChain)) => {}
+                        Err(BlockImportError::Import(ImportError::AlreadyInChain)) => {
+                            self.state = State::SnapshotChunk(*header.state_root());
+                        }
                         Err(BlockImportError::Import(ImportError::AlreadyQueued)) => {}
                         // FIXME: handle import errors
                         Err(err) => {
                             cwarn!(SYNC, "Cannot import header({}): {:?}", header.hash(), err);
-                            break
                         }
                         _ => {}
                     }
                 }
-            }
+                _ => cdebug!(SYNC, "Peer {} didn't have the snapshot header {}", from, hash),
```

Comment:
The log message is misleading. inversion of `[header] if header.hash() == hash` is "headers is not `[header]`" (such as `[]`, `[h1, ...]`), or `header.hash() != hash`. You should change it to `"Peer {} responded with the invalid response. requested hash: {}, response  length: {}"`

Reply:
I thought it is checked in `is_valid_response`, but it seems it's not.
I'll fix this.